### ISSUE TITLE
Support custom form controls checkbox, radio and switch

### ISSUE THREAD

**Issue by jelhan** [kaliber5/ember-bootstrap#912]
Bootstrap 4 supports advanced checkboxes and radio controls. First one could also be styled as a switch. Ember Bootstrap should support them as well. If I didn't missed something it's only requires different CSS classes.

Bootstrap docs:
- Checkboxes and radios: https://getbootstrap.com/docs/4.3/components/forms/#checkboxes-and-radios-1
- Switches: https://getbootstrap.com/docs/4.3/components/forms/#switches

**Comment by adambedford** [comment]
I'd love to see this!

**Comment by jelhan** [comment]
@adambedford Let me know if you have time to work on this one. Happy to help you with any question.

**Comment by lindyhopchris** [comment]
@jelhan hope you're ok! was wondering what needs to happen to get this progressed? I really need to use the switches, so I'd be prepared to do a PR. I just need some guidance as to the expectation for how this should be implemented, so that I don't waste any of my limited time on the PR?

Using switches as an example, is the intent to add a new control type? Or should "checkbox" still be used but some other argument control the custom control type?

**Comment by jelhan** [comment]
Thanks a lot for taking this one up.

> Using switches as an example, is the intent to add a new control type? Or should "checkbox" still be used but some other argument control the custom control type?

I think we have two options for public API:

```
<BsForm as |form|>
  <form.element @controlType="switch" />
</BsForm>
````

and

````
<BsForm as |form|>
  <form.element @controlType="checkbox" as |element|>
    <element.control @style="switch" />
  </form.element>
</BsForm>
```

I haven't evaluated if both styles are supported by current architecture. It depends on the HTML markup expected by Bootstrap. But I hope both are.

To evaluate if you could achieve (and to have a quick work-a-round) you could try to construct HTML markup expected by Bootstrap inside of block yielded by `<form.element>`. If this isn't possible, we need to talk. Otherwise it could be implemented as a flavor of `checkbox` control.

If possible I would tend towards controlling by an argument of `checkbox` control. Switch and checkbox seems to be a visual difference only. It's not another input concept. But I'm not very opinionated about it.

**Comment by lindyhopchris** [comment]
So the problem is the wrapping `<div>`

A standard checkbox wraps the input in a `<div class="form-check">`.

The custom checkbox wraps the input in `<div class="custom-control custom-checkbox">` and the switch uses `<div class="custom-control custom-switch">`

That means the `<element.control @style="switch" />` won't work, because that's setting the style on the `<input>` element, when the wrapping `<div>` needs to be modified as well as the `<input>`.

So my guess would be we need to go for a `@controlType="switch"` approach. But my question on that is what should all the control types be called? I think maybe `custom-checkbox`, `custom-radio`, `custom-switch` as that matches what Bootstrap calls them?

**Comment by jelhan** [comment]
> So the problem is the wrapping `<div>`
> 
> A standard checkbox wraps the input in a `<div class="form-check">`.
> 
> The custom checkbox wraps the input in `<div class="custom-control custom-checkbox">` and the switch uses `<div class="custom-control custom-switch">`
>
> That means the `<element.control @style="switch" />` won't work, because that's setting the style on the `<input>` element, when the wrapping `<div>` needs to be modified as well as the `<input>`.

Not being able to implement support for this as a control seems to highlight an architectural issue. It is the [same for Bootstrap v5](https://getbootstrap.com/docs/5.0/forms/checks-radios/#switches). So I think we should fix in long term.

> So my guess would be we need to go for a `@controlType="switch"` approach. But my question on that is what should all the control types be called? I think maybe `custom-checkbox`, `custom-radio`, `custom-switch` as that matches what Bootstrap calls them?

I'm tending towards `@controlType="switch"`. It feels more inline with the other values. `custom-switch` opens up the question what is _custom_ about it.

I see that we need to have `custom-checkbox` and `custom-radio` because `checkbox` and `radio` are already taken. I had a quick look in v5 documentation and wasn't able to find custom checkbox and custom radio mentioned in it. Not sure if I missed something. But otherwise we may not need them for long-term. This would make it easier to accept `@controlType="custom-checkbox"` but `@controlType="switch"`.

@simonihmig Any thoughts?

**Comment by lindyhopchris** [comment]
Yeah you're right that I can only see the switch in BS5. So probably the only custom control we want to implement in BS4 would be the switch, as the others wouldn't transfer over to v5. So yeah, maybe `@controlType="switch"` is the best way to go.

**Comment by simonihmig** [comment]
> Personally I'm not sure it's worth adding the custom checkboxes and radios at this stage, not least because they don't actually exist in BS5. But that's just my opinion!

Yeah, I agree! That's why I asked. :)

> I'll do a PR just to add the switch control type.

That would be awesome! Feel free to reach out anytime if you need!

**Comment by jelhan** [comment]
You are right. I only looked at switch.
